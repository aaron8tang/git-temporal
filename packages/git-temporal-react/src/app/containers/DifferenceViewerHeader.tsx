import React from 'react';
// @ts-ignore
import { useSelector, useDispatch, shallowEqual } from 'react-redux';

import { style } from 'app/styles';

import {
  getStartDate,
  getEndDate,
  getDiffLeftCommit,
  getDiffRightCommit,
  getSelectedPath,
} from 'app/selectors/stateVars';
import {
  getCommitsForTimeplot,
  getFilteredCommits,
} from 'app/selectors/commits';
import { setDates } from 'app/actions/setDates';

import { RevSelector } from 'app/components/RevSelector';
import { EpochDateTime } from 'app/components/EpochDateTime';
import { fetchDiff } from 'app/actions/diff';

const outerStyle = {
  _extends: ['flexRow'],
  flexGrow: 0,
  flexShrink: 0,
  background: '@colors.altBackground',
  color: '@colors.altForeground',
};

const revChildrenStyle = {
  _extends: 'flexColumn',
  alignItems: 'center',
  minWidth: 160,
};

const revSelectorStyle = {
  flexGrow: 1,
  textAlign: 'center',
};

export const DifferenceViewerHeader: React.FC = (): React.ReactElement => {
  const selectedPath = useSelector(getSelectedPath);
  const startDate = useSelector(getStartDate);
  const endDate = useSelector(getEndDate);
  const leftCommit = useSelector(getDiffLeftCommit);
  const rightCommit = useSelector(getDiffRightCommit);
  const timeplotCommits = useSelector(getCommitsForTimeplot);
  const dispatch = useDispatch();

  return (
    <div style={style(outerStyle)}>
      <RevSelector
        style={style(revSelectorStyle)}
        disablePrevious={shouldDisablePrevious(leftCommit)}
        disableNext={shouldDisableNext(leftCommit)}
        onNextRevClick={onLeftRevNext}
        onPreviousRevClick={onLeftRevPrevious}
      >
        {renderRevChildren('left', leftCommit)}
      </RevSelector>
      <RevSelector
        style={style(revSelectorStyle)}
        disablePrevious={shouldDisablePrevious(rightCommit)}
        disableNext={shouldDisableNext(rightCommit)}
        onNextRevClick={onRightRevNext}
        onPreviousRevClick={onRightRevPrevious}
      >
        {renderRevChildren('right', rightCommit)}
      </RevSelector>
    </div>
  );

  function renderRevChildren(which, _commit) {
    if (!timeplotCommits || timeplotCommits.length === 0) {
      return null;
    }
    const commit = which === 'left' && !_commit ? timeplotCommits[0] : _commit;

    return (
      <div style={style(revChildrenStyle)}>
        {commit ? (
          <>
            <div style={style('flexRow')}>
              <EpochDateTime value={commit.authorDate} />
            </div>
            <div style={style('flexRow')}>
              (#
              {commit.hash})
              {commit.id === timeplotCommits[0].id && ' (Local HEAD)'}
            </div>
          </>
        ) : (
          <div style={style('flexRow')}>Uncommitted Changes</div>
        )}
      </div>
    );
  }

  function shouldDisablePrevious(commit) {
    if (!timeplotCommits) {
      return true;
    }
    // a null left or right commit means it's the local changes
    // or the HEAD rev and there should be previous commits
    return (
      (commit && commit === timeplotCommits[timeplotCommits.length - 1]) ||
      (commit && commit === rightCommit && rightCommit === leftCommit)
    );
  }
  function shouldDisableNext(commit) {
    // a null commit means it's the latest rev
    if (!timeplotCommits || !commit) {
      return true;
    }
    return commit === null;
  }

  function onLeftRevPrevious() {
    setLeftRev(-1);
  }

  function onLeftRevNext() {
    setLeftRev(1);
  }

  function setLeftRev(relativeIndex) {
    if (!timeplotCommits || timeplotCommits.length === 0) {
      return;
    }
    let newLeftCommit = null;
    if (leftCommit) {
      const foundIndex = timeplotCommits.findIndex(commit => {
        return commit.id === leftCommit.id;
      });
      if (foundIndex !== -1) {
        newLeftCommit = timeplotCommits[foundIndex - relativeIndex];
      }
    } else if (relativeIndex < 0) {
      newLeftCommit = timeplotCommits[2];
    }
    if (newLeftCommit) {
      dispatch(setDates((newLeftCommit.authorDate + 1) * 1000, endDate * 1000));
    }
  }

  function onRightRevPrevious() {
    setRightRev(-1);
  }

  function onRightRevNext() {
    setRightRev(1);
  }

  function setRightRev(relativeIndex) {
    if (!timeplotCommits || timeplotCommits.length === 0) {
      return;
    }
    let newRightCommit = null;
    if (rightCommit) {
      const foundIndex = timeplotCommits.findIndex(commit => {
        return commit.id === rightCommit.id;
      });
      if (foundIndex !== -1) {
        newRightCommit = timeplotCommits[foundIndex - relativeIndex];
      }
    } else if (relativeIndex < 0) {
      newRightCommit = timeplotCommits[0];
    }
    if (newRightCommit) {
      dispatch(
        setDates(startDate * 1000, (newRightCommit.authorDate + 1) * 1000)
      );
    }
  }
};
