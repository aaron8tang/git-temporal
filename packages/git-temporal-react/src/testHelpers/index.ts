import reduxMockStore from 'redux-mock-store';
import reduxThunk from 'redux-thunk';

const middlewares = [reduxThunk]; // add your middlewares like `redux-thunk`

const defaultInitialState = {
  commits: [],
  selectedPath: '',
};

export function mockReduxStore() {
  return reduxMockStore(middlewares)(defaultInitialState);
}
