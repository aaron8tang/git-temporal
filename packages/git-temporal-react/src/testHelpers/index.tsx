import React from 'react';
import { render } from '@testing-library/react';
import { createStore, applyMiddleware, compose } from 'redux';
import { Provider } from 'react-redux';
import reduxThunk from 'redux-thunk';

import reducers from '../app/reducers';
import { ReduxStoreMock } from './mocks/reduxStore';

const middleware = [reduxThunk]; // add your middlewares like `redux-thunk`

export function mockReduxStore(initialStore = {}) {
  return createStore(
    reducers,
    // @ts-ignore
    { ...ReduxStoreMock, ...initialStore },
    compose(applyMiddleware(...middleware))
  );
}

export function mountConnected(children, initialStore = {}) {
  const store = mockReduxStore(initialStore);
  jest.spyOn(store, 'dispatch');

  const wrapper = render(<Provider store={store}>{children}</Provider>);
  return { store, wrapper };
}
