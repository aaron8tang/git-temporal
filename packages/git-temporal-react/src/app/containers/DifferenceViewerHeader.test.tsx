import React from 'react';
import { mountConnected } from 'testHelpers';

import { DifferenceViewerHeader } from './DifferenceViewerHeader';

describe('containers/DifferenceViewerHeader', () => {
  describe('when rendered with mock redux store', () => {
    it('should match snapshot', () => {
      const { wrapper } = mountConnected(<DifferenceViewerHeader />);
      expect(wrapper.baseElement).toMatchSnapshot();
    });
    it('should render all of the things', async () => {
      const { wrapper } = mountConnected(<DifferenceViewerHeader />);

      // from redux mock store data, we should see....
      wrapper.getByText('Jan 12, 2020, 4 AM');
      wrapper.getByText('(#94932bed) (Local HEAD)');
    });
  });
});
