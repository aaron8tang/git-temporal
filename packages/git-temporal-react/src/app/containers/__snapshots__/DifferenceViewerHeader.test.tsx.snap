// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/DifferenceViewerHeader when rendered with mock redux store should match snapshot 1`] = `
<body>
  <div>
    <div
      style="display: flex; flex-direction: row; flex-grow: 0; flex-shrink: 0; background: whitesmoke; color: rgb(51, 51, 51);"
    >
      <div
        style="display: flex; flex-direction: row; font-size: 12px; line-height: 1.35; font-weight: normal; align-self: baseline; flex-grow: 1; text-align: center;"
      >
        <div
          style="flex-grow: 1;"
        />
        <div
          style="font-size: 12px; line-height: 1.35; font-weight: normal; padding: 5px; border: 1px solid transparent; display: flex; flex-direction: row; color: rgb(51, 51, 51); margin-bottom: 5px;"
        >
          <svg
            height="16"
            role="img"
            viewBox="0 0 192 512"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
              fill="currentColor"
            />
          </svg>
        </div>
        <div
          style="display: flex; flex-direction: column; flex-grow: 0; align-items: center;"
        >
          <div
            style="display: flex; flex-direction: column; align-items: center; min-width: 160px;"
          >
            <div
              style="display: flex; flex-direction: row;"
            >
              <span>
                Dec 30, 2019, 11 AM
              </span>
            </div>
            <div
              style="display: flex; flex-direction: row;"
            >
              (#
              c03bb366
              )
            </div>
          </div>
        </div>
        <div
          style="font-size: 12px; line-height: 1.35; font-weight: normal; padding: 5px; border: 1px solid transparent; display: flex; flex-direction: row; color: rgb(51, 51, 51); margin-bottom: 5px;"
        >
          <svg
            height="16"
            role="img"
            viewBox="0 0 192 512"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
              fill="currentColor"
            />
          </svg>
        </div>
        <div
          style="flex-grow: 1;"
        />
      </div>
      <div
        style="display: flex; flex-direction: row; font-size: 12px; line-height: 1.35; font-weight: normal; align-self: baseline; flex-grow: 1; text-align: center;"
      >
        <div
          style="flex-grow: 1;"
        />
        <div
          style="font-size: 12px; line-height: 1.35; font-weight: normal; padding: 5px; border: 1px solid transparent; display: flex; flex-direction: row; color: rgb(51, 51, 51); margin-bottom: 5px;"
        >
          <svg
            height="16"
            role="img"
            viewBox="0 0 192 512"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
              fill="currentColor"
            />
          </svg>
        </div>
        <div
          style="display: flex; flex-direction: column; flex-grow: 0; align-items: center;"
        >
          <div
            style="display: flex; flex-direction: column; align-items: center; min-width: 160px;"
          >
            <div
              style="display: flex; flex-direction: row;"
            >
              <span>
                Jan 12, 2020, 4 AM
              </span>
            </div>
            <div
              style="display: flex; flex-direction: row;"
            >
              (#
              94932bed
              )
               (Local HEAD)
            </div>
          </div>
        </div>
        <div
          style="font-size: 12px; line-height: 1.35; font-weight: normal; padding: 5px; border: 1px solid transparent; display: flex; flex-direction: row; color: rgb(51, 51, 51); margin-bottom: 5px;"
        >
          <svg
            height="16"
            role="img"
            viewBox="0 0 192 512"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
              fill="currentColor"
            />
          </svg>
        </div>
        <div
          style="flex-grow: 1;"
        />
      </div>
    </div>
  </div>
</body>
`;
